### PR TITLE
`Lectures`: Allow instructors to define relations between learning goals

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German. _(Did not translate admin-only interface.)_

### Motivation and Context

Working towards integrating Adaptive Learning in Artemis, instructors need to be able to define relations between learning goals. This will later allow Artemis to generate Learning Paths between the individual learning goals.

### Description

We model a new relation between two learning goals that describes their relation on a graph-like structure (you can imagine directed edges in a graph.). Additionally, a new attribute is added to learning goals to persist their [taxonomy](https://en.wikipedia.org/wiki/Bloom%27s_taxonomy).
To make this as versatile as possible, the relation is defined through a new class called `LearningGoalRelation` that holds the tail learning goal, head learning goal and the relation type:

- RELATES: A generic relation between two learning goals.
- ASSUMES: The tail learning goal assumes that the student already achieved the head learning goal.
- EXTENDS: The tail learning goal extends the head learning goal on the same topic in more detail.
- MATCHES: The tail learning goal matches the head learning goal (e.g., a duplicate).

To allow you to test this feature, we implemented an admin-only interface with which you to create relations, display them in a graph and remove them again (by clicking on them in the graph).

### Steps for Testing

Prerequisites:
- 1 Admin

1. Log in to Artemis as admin
2. Navigate to Course Administration
3. Go to the Learning Goal Management page
4. Create some learning goals
5. Add some relations between the learning goals and make sure they are persisted
7. Remove a relation by clicking on it in the graph

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Method | Line |
|------------|-------:|-----:|
| LearningGoalService.java | 100% | 96% |
| LearningGoalResource.java | 100% | 91% |

### Screenshots

Screenshot 2022-08-30 at 17 21 25

### CODE REVIEW

**Comment by krusche** [comment]
there is no feature for Learning goals (yet), so I renamed it to lectures as it currently is mostly related to lecture units. I know we change this in the future, but let's stick (for now) to the features defined in the README on Github and about page on Artemis.

I am happy to add a new feature such as "Competency-based learning" in the future, but we should first add it to the README on Github and the about page (with a minimal documentation) before using it.

File: src/main/java/de/tum/in/www1/artemis/domain/LearningGoalRelation.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        CONSECUTIVE;
+    }
+
+    @Converter
+    public static class RelationTypeConverter implements AttributeConverter<RelationType, String> {
+
+        @Override
+        public String convertToDatabaseColumn(RelationType type) {
+            if (type == null)
+                return null;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/domain/LearningGoalRelation.java:70]: Codacy

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java [hunk 2]
```diff
@@ -276,6 +300,55 @@ public ResponseEntity<SearchResultPageDTO<LearningGoal>> getAllLecturesOnPage(Pa
...
+     * POST /courses/:courseId/goals/:learningGoalId/relations
+     * @param courseId  the id of the course to which the learning goals belongs
+     * @param tailLearningGoalId the id of the prerequisite (learning goal) to add
+     * @param headLearningGoalId the id of the prerequisite (learning goal) to add
+     * @return the ResponseEntity with status 200 (OK)
+     */
+    @PostMapping("/courses/{courseId}/goals/{tailLearningGoalId}/relations/{headLearningGoalId}")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<LearningGoalRelation> createLearningGoalRelation(@PathVariable Long courseId, @PathVariable Long tailLearningGoalId,
+            @PathVariable Long headLearningGoalId, @RequestParam(defaultValue = "") String type) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java:313]: Codacy

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java [hunk 3]
```diff
@@ -276,6 +300,55 @@ public ResponseEntity<SearchResultPageDTO<LearningGoal>> getAllLecturesOnPage(Pa
...
+     * DELETE /courses/:courseId/goals/:learningGoalId/relations/:learningGoalRelationId
+     * @param courseId the id of the course for which the learning goal is a prerequisite
+     * @param learningGoalId the id of the prerequisite (learning goal) to remove
+     * @return the ResponseEntity with status 200 (OK)
+     */
+    @DeleteMapping("/courses/{courseId}/goals/{learningGoalId}/relations/{learningGoalRelationId}")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<Void> removeLearningGoalRelation(@PathVariable Long learningGoalId, @PathVariable Long courseId, @PathVariable Long learningGoalRelationId) {
+        log.info("REST request to remove a learning goal relation: {}", learningGoalId);
+        var learningGoal = findLearningGoal(Role.INSTRUCTOR, learningGoalId, courseId, false, false);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java:344]: Codacy

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java [hunk 4]
```diff
@@ -276,6 +300,55 @@ public ResponseEntity<SearchResultPageDTO<LearningGoal>> getAllLecturesOnPage(Pa
...
+
+    /**
+     * DELETE /courses/:courseId/goals/:learningGoalId/relations/:learningGoalRelationId
+     * @param courseId the id of the course for which the learning goal is a prerequisite
+     * @param learningGoalId the id of the prerequisite (learning goal) to remove
+     * @return the ResponseEntity with status 200 (OK)
+     */
+    @DeleteMapping("/courses/{courseId}/goals/{learningGoalId}/relations/{learningGoalRelationId}")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    public ResponseEntity<Void> removeLearningGoalRelation(@PathVariable Long learningGoalId, @PathVariable Long courseId, @PathVariable Long learningGoalRelationId) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LearningGoalResource.java:342]: Codacy

File: src/main/webapp/i18n/de/learningGoal.json [hunk 5]
```diff
@@ -7,6 +7,15 @@
             "learningGoalButton": "Lernziele",
             "title": "Lernziele",
             "description": "Beschreibung",
+            "taxonomy": "Taxonomy",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/learningGoal.json:10]:
```suggestion
            "taxonomy": "Taxonomie",
```
Is there no better German word?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by maximiliansoelch — DISMISSED** [review]
Tested locally, works as expected 👍

**Review by akesfeden — COMMENTED** [review]
Tested locally, overall works as described, nice feature! I noticed that in this configuration the diagram is not fit to view properly on first load, so addressing may be nice. But since it is draggable it is not very critical or a blocking issue, what do you think?

Also, it may be better in terms of usability if the diagram is updated in-place when a change is made so that the user does not need to expand it everytime, but again, this is not very urgent.

image

**Review by stefanwaldhauser — APPROVED** [review]
Tested locally! Works great and code looks good too!

**Review by stefanwaldhauser — APPROVED** [review]
Reapprove! :)

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
You have some Codacy issues in [learning-goal-management.component.spec.ts](https://github.com/ls1intum/Artemis/pull/5311/files#diff-abe6e1859ae3e9f533cfd13bd67f3795cc615cac00bf54ec9b1f08e7977e03ce)

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me

File: src/test/java/de/tum/in/www1/artemis/LearningGoalIntegrationTest.java [hunk 6]
```diff
@@ -408,6 +436,70 @@ void deleteCourse_asAdmin_shouldAlsoDeleteLearningGoal() throws Exception {
...
+        Long idOfOtherLearningGoal = database.createLearningGoal(course).getId();
+
+        request.post(
+                "/api/courses/" + idOfCourse + "/goals/" + idOfLearningGoal + "/relations/" + idOfOtherLearningGoal + "?type=" + LearningGoalRelation.RelationType.EXTENDS.name(),
+                null, HttpStatus.FORBIDDEN);
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void gLearningGoalRelations() throws Exception {
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/LearningGoalIntegrationTest.java:467]:
```suggestion
    public void getLearningGoalRelations() throws Exception {
```

I assume it should be like this

File: src/test/java/de/tum/in/www1/artemis/LearningGoalIntegrationTest.java [hunk 7]
```diff
@@ -395,6 +407,22 @@ void deleteLearningGoal_asInstructor_shouldDeleteLearningGoal() throws Exception
         request.get("/api/courses/" + idOfCourse + "/goals/" + idOfLearningGoal, HttpStatus.NOT_FOUND, LearningGoal.class);
     }
 
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void deleteLearningGoal_witRelatedGoals_shouldReturnBadRequest() throws Exception {
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/LearningGoalIntegrationTest.java:412]: Optional request: Maybe add a comment, why this test case returns a bad request (what condition was not fulfilled)

**Review by sleiss — DISMISSED** [review]
Tested it in the testing session, code also looks good to me :+1:

**Review by Dominik-Weinzierl — APPROVED** [review]
Code looks good to me - clean structure 👍 in addition, i like the usage of the converter 🚀

**Review by dangcosta — APPROVED** [review]
Tested locally and works as expected.

**Review by akesfeden — APPROVED** [review]
Tested again locally, works as described. My points in the previous comment https://github.com/ls1intum/Artemis/pull/5311#pullrequestreview-1092269870 still hold, maybe they can be done in follow-ups, what do you think?

**Comment by ge65cer** [comment]
@akesfeden Yes, I've noted it down. 🙂

**Review by Dominik-Weinzierl — APPROVED** [review]
Reapprove - looks still good

**Review by manuelmanso — APPROVED** [review]
Re-approve - thanks for considering the feedback

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove!

**Review by sleiss — DISMISSED** [review]
Re-approve

**Pull request merged**